src/Ztm.WebApi/AddressPools/ReceivingAddressPool.cs

```diff
@@ -0,0 +1,120 @@
+            IReceivingAddressStorage storage,
+            IAddressChoser choser)
+        {
+            if (generator == null)
+            {
+                throw new ArgumentNullException(nameof(generator));
+            }
+
+            if (storage == null)
+            {
+                throw new ArgumentNullException(nameof(storage));
+            }
+
+            if (choser == null)
+            {
+                throw new ArgumentNullException(nameof(choser));
+            }
+
+            this.generator = generator;
+            this.storage = storage;
+            this.choser = choser;
+
+            this.storageLock = new ReaderWriterLockSlim();
+        }
+
+        public async Task GenerateAddressAsync(CancellationToken cancellationToken)
+        {
+            storageLock.EnterWriteLock();
+
+            try
+            {
+                var address = await this.generator.GenerateAsync(cancellationToken);
+                await this.storage.AddAddressAsync(address, cancellationToken);
+            }
+            finally
+            {
+                storageLock.ExitWriteLock();
+            }
+        }
+
+        public async Task ReleaseAsync(Guid id, CancellationToken cancellationToken)
+        {
+            this.storageLock.EnterWriteLock();
+
+            try
+            {
+                var r = await this.storage.GetReservationAsync(id, cancellationToken);
+                if (r == null)
+                {
+                    throw new KeyNotFoundException("The reservation is not found.");
+                }
+
+                if (r.ReleasedDate != null)
+                {
+                    throw new InvalidOperationException("The reservation is already released.");
+                }
+
+                await this.storage.SetLockedStatusAsync(r.ReceivingAddress.Id, false, cancellationToken);
+                await this.storage.SetReleasedTimeAsync(id, cancellationToken);
+            }
+            finally
+            {
+                this.storageLock.ExitWriteLock();
+            }
+        }
+
+        public async Task<ReceivingAddressReservation> TryLockAddressAsync(CancellationToken cancellationToken)
+        {
+            this.storageLock.EnterWriteLock();
+
+            try
+            {
+                var addresses = await this.storage.ListReceivingAddressAsync(cancellationToken);
+                if (addresses != null)
+                {
+                    var availables = addresses.Where(a => a.Available);
+
+                    if (availables.Count() <= 0)
+                    {
+                        return null;
+                    }
+
+                    var chosen = this.choser.Choose(availables);
+
+                    if (chosen != null)
+                    {
+                        var reservation = await this.storage.CreateReservationAsync(chosen.Id, cancellationToken);
+                        await this.storage.SetLockedStatusAsync(chosen.Id, true, cancellationToken);
```

Comment:
The locked state should be combined together with reservation creation and must be atomic operation.